Source/Internal/UICollectionViewLayout+InteractiveReordering.m

```diff
@@ -0,0 +1,202 @@
+#import <IGListKit/IGListAdapterInternal.h>
+#import <IGListKit/IGListAssert.h>
+#import <IGListKit/IGListSectionController.h>
+
+#import <objc/runtime.h>
+
+@implementation UICollectionViewLayout (InteractiveReordering)
+
+static void * kIGListAdapterKey = &kIGListAdapterKey;
+
++ (void)load
+{
+    static dispatch_once_t onceToken;
+    dispatch_once(&onceToken, ^{
+        // interactive reordering does not exist prior to iOS 9
+        if ([[[UIDevice currentDevice] systemVersion] floatValue] < 9.0) {
+            return;
+        }
+
+        Class layoutClass = [self class];
+
+        // override implementation for targetIndexPathForInteractivelyMovingItem:withPosition:
+        SEL userMoveSelector = @selector(targetIndexPathForInteractivelyMovingItem:withPosition:);
+        SEL overrideSelector = @selector(ig_targetIndexPathForInteractivelyMovingItem:withPosition:);
+        Method userLayoutMethod = class_getInstanceMethod(layoutClass, userMoveSelector);
+        Method overrideLayoutMethod = class_getInstanceMethod(layoutClass, overrideSelector);
+        method_exchangeImplementations(userLayoutMethod, overrideLayoutMethod);
+
+        // override implementation for
+        // invalidationContextForInteractivelyMovingItems:withTargetPosition:previousIndexPaths:previousPosition:
+        SEL userInvalidationSelector =
+        @selector(invalidationContextForInteractivelyMovingItems:withTargetPosition:previousIndexPaths:previousPosition:);
+        SEL overrideInvalidationSelector =
+        @selector(ig_invalidationContextForInteractivelyMovingItems:withTargetPosition:previousIndexPaths:previousPosition:);
+        Method userInvalidationMethod = class_getInstanceMethod(layoutClass, userInvalidationSelector);
+        Method overrideInvalidationMethod = class_getInstanceMethod(layoutClass, overrideInvalidationSelector);
+        method_exchangeImplementations(userInvalidationMethod, overrideInvalidationMethod);
+
+        // override implementation for
+        // invalidationContextForInteractivelyMovingItems:withTargetPosition:previousIndexPaths:previousPosition:
+        SEL userEndInvalidationSelector =
+        @selector(invalidationContextForEndingInteractiveMovementOfItemsToFinalIndexPaths:previousIndexPaths:movementCancelled:);
+        SEL overrideEndInvalidationSelector =
+        @selector(ig_invalidationContextForEndingInteractiveMovementOfItemsToFinalIndexPaths:previousIndexPaths:movementCancelled:);
+        Method userEndInvalidationMethod = class_getInstanceMethod(layoutClass, userEndInvalidationSelector);
+        Method overrideEndInvalidationMethod = class_getInstanceMethod(layoutClass, overrideEndInvalidationSelector);
+        method_exchangeImplementations(userEndInvalidationMethod, overrideEndInvalidationMethod);
+    });
+}
+
+- (void)ig_hijackLayoutInteractiveReorderingMethodForAdapter:(IGListAdapter *)adapter {
+    objc_setAssociatedObject(self, kIGListAdapterKey, adapter, OBJC_ASSOCIATION_ASSIGN);
```

Comment:
This may have been a misread on my part. I'm not seeing cycles in the memory debugger or leaks in instruments, but with retain, I'm failing 5 tests that all seem to be checking that the adapter gets released. Ex. `test_whenReleasingObjects_thatAssertDoesntFire` 🤔

Reply:
Oh! I see.